# Apply the Artificial Neural Network to Logistic Regression for Pattern Data Set

## Train and Test the ANN by splitting the CSV dataset and using a single Logistic Regression layer

In [36]:
# Import numpy
import numpy as np

# Unity/Identity activation function for layer 0
def identity_activation(G0):
    return G0

# TanH activation function for layer 1
def tanh_activation(G1):
    return (np.exp(G1)-np.exp(-G1))/(np.exp(G1)+np.exp(-G1))

# Sigmoid activation function for layer 2
def sigmoid_activation(G2):
    return 1/(1+np.exp(-G2))

# Initialize Weights and Biases
def initialize_weights_and_biases(n):
    W0 = np.identity(n)
    b0 = np.zeros((n,1))
    
    # Replaced identity matrix with a matrix that reproduces two of four outputs exactly as inputs and the other two as zeros
    W1 = np.array([[1,0],[0,0],[0,1],[0,0]])
    b1 = np.zeros((2*n,1))
    
    W2 = np.ones((1,2*n))
    b2 = np.ones((1,1))
    
    return W0, b0, W1, b1, W2, b2

# Cost function
def cost_function(Y, Yhat, m):
    return -(1/m)*np.sum(Y*np.log(Yhat) + (1-Y)*np.log(1-Yhat))

def ann_logit(filename, n_train, n_features):
    
    # Read CSV data from input file
    file_csv = np.genfromtxt (filename, delimiter=",")

    # Get the Training datasets
    X_train = file_csv[1:n_train+1,0:n_features].transpose()
    Y_train = file_csv[1:n_train+1,n_features:n_features+1].transpose()

    # Get the Testing datasets
    X_test = file_csv[n_train+1:,0:n_features].transpose()
    Y_test = file_csv[n_train+1:,n_features:n_features+1].transpose()

    # Number of training examples
    m = X_train.shape[1]

    # Number of training features
    n = X_train.shape[0]

    # Iterations
    h = 100000

    # Learning rate
    alpha = 0.05

    # Initialize the Weights and Biases
    W0, b0, W1, b1, W2, b2 = initialize_weights_and_biases(n)

    # Initialize the J cost function array
    J = np.zeros((1,h))

    # Loop through the forward and backward steps
    for g in range(h):

        # Forward Propagation through the layers 0, 1 and 2
        # Layer 0 - Unity
        X1 = identity_activation(W0@X_train + b0)

        # Layer 1 - Unity
        X2 = identity_activation(W1@X1 + b1)

        # Layer 2 - Sigmoid
        Yhat = sigmoid_activation(W2@X2 + b2)

        # Calculate the Cost Function
        J[0][g] = cost_function(Y_train, Yhat, m)

        # Use derivatives of J cost funtion to update the Weights and Biases
        dJdz = Yhat - Y_train
        dJdb = (1/m)*np.sum(dJdz)
        dJdw = (1/m)*dJdz@np.transpose(X2)
        b2 = b2 - alpha*dJdb
        W2 = W2 - alpha*dJdw

    print("The Weights array after training is: {}".format(W2))
    print("The Biases array after training is: {}".format(b2))
    print("The last value of the Cost Function is: {}".format(J[0][h-1]))

    ## Test the parms with the testing data
    # Apply the model parameters derived during training with the sigmoid function in the activation routine
    # Forward Propagation through the layers 0, 1 and 2
    # Layer 0 - Unity
    X1 = identity_activation(W0@X_test + b0)

    # Layer 1 - Unity
    X2 = identity_activation(W1@X1 + b1)

    # Layer 2 - Sigmoid
    Yhat = sigmoid_activation(W2@X2 + b2)
    
    # Round the Yhat values to 1 or 0
    Yhat_test = np.rint(Yhat)

    # Compare the Yhat values to the Y values from the Testing data
    Ycomp_test = np.equal(Y_test,Yhat_test)
    
    # Print out the accuracy using the mean of True/False values
    print("The percentage accuracy from 320 training examples and 80 testing examples is: {}%".format(np.mean(Ycomp_test)*100))
    
# Call the ANN logitistic regression for the training data set
# Pass parms:: filename, number of examples to use for training, number of X features in the data
ann_logit('pattern_rand.csv', 320, 2)


The Weights array after training is: [[ 0.08056186  1.         -0.24098445  1.        ]]
The Biases array after training is: [[-0.00078901]]
The last value of the Cost Function is: 0.6636268397341353
The percentage accuracy from 320 training examples and 80 testing examples is: 45.0%


The Weights array after training is: [[ 0.08056186  1.         -0.24098445  1.        ]]

The Biases array after training is: [[-0.00078901]]

The last value of the Cost Function is: 0.6636268397341353

The percentage accuracy from 320 training examples and 80 testing examples is: 45.0%